File: aegismatrix-engine/shared/schema.ts
import { sql } from "drizzle-orm";
import { pgTable, text, varchar, real, integer, jsonb, timestamp } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod";

// Market Data Table
export const marketData = pgTable("market_data", {
  id: varchar("id").primaryKey().default(sql`gen_random_uuid()`),
  spot: real("spot").notNull(),
  spotChange: real("spot_change").notNull(),
  vix: real("vix").notNull(),
  vixChange: real("vix_change").notNull(),
  regime: text("regime").notNull(),
  generatedAt: timestamp("generated_at").notNull().defaultNow(),
});

export const insertMarketDataSchema = createInsertSchema(marketData).omit({ id: true, generatedAt: true });
export type InsertMarketData = z.infer<typeof insertMarketDataSchema>;
export type MarketData = typeof marketData.$inferSelect;

// Direction Horizon Table
export const directionHorizon = pgTable("direction_horizon", {
  id: varchar("id").primaryKey().default(sql`gen_random_uuid()`),
  marketDataId: varchar("market_data_id").notNull(),
  horizonType: text("horizon_type").notNull(), // "today", "t1", "t3", etc.
  direction: text("direction").notNull(),
  expectedMovePoints: real("expected_move_points").notNull(),
  conviction: real("conviction").notNull(),
  generatedAt: timestamp("generated_at").notNull().defaultNow(),
});

export const insertDirectionHorizonSchema = createInsertSchema(directionHorizon).omit({ id: true, generatedAt: true });
export type InsertDirectionHorizon = z.infer<typeof insertDirectionHorizonSchema>;
export type DirectionHorizon = typeof directionHorizon.$inferSelect;

// Direction Risk Table
export const directionRisk = pgTable("direction_risk", {
  id: varchar("id").primaryKey().default(sql`gen_random_uuid()`),
  marketDataId: varchar("market_data_id").notNull(),
  riskScore: real("risk_score").notNull(),
  generatedAt: timestamp("generated_at").notNull().defaultNow(),
});

export const insertDirectionRiskSchema = createInsertSchema(directionRisk).omit({ id: true, generatedAt: true });
export type InsertDirectionRisk = z.infer<typeof insertDirectionRiskSchema>;
export type DirectionRisk = typeof directionRisk.$inferSelect;

// Seller Data Table
export const sellerData = pgTable("seller_data", {
  id: varchar("id").primaryKey().default(sql`gen_random_uuid()`),
  marketDataId: varchar("market_data_id").notNull(),
  safeLower: real("safe_lower").notNull(),
  safeUpper: real("safe_upper").notNull(),
  maxPainLower: real("max_pain_lower").notNull(),
  maxPainUpper: real("max_pain_upper").notNull(),
  expiryStressScore: real("expiry_stress_score").notNull(),
  expiryStressLevel: text("expiry_stress_level").notNull(),
  trapScore: real("trap_score").notNull(),
  trapLevel: text("trap_level").notNull(),
  putSkew: real("put_skew").notNull(),
  callSkew: real("call_skew").notNull(),
  regime: text("regime").notNull(),
  historicalHitRate: real("historical_hit_rate").notNull(),
  dailyFlag: text("daily_flag").notNull(),
  breachProbabilities: jsonb("breach_probabilities").notNull(),
  generatedAt: timestamp("generated_at").notNull().defaultNow(),
});

export const insertSellerDataSchema = createInsertSchema(sellerData).omit({ id: true, generatedAt: true });
export type InsertSellerData = z.infer<typeof insertSellerDataSchema>;
export type SellerData = typeof sellerData.$inferSelect;

// Buyer Data Table
export const buyerData = pgTable("buyer_data", {
  id: varchar("id").primaryKey().default(sql`gen_random_uuid()`),
  marketDataId: varchar("market_data_id").notNull(),
  breakoutTodayScore: real("breakout_today_score").notNull(),
  breakoutTodayLevel: text("breakout_today_level").notNull(),
  spikeUpProb: real("spike_up_prob").notNull(),
  spikeDownProb: real("spike_down_prob").notNull(),
  thetaEdgeScore: real("theta_edge_score").notNull(),
  thetaEdgeLevel: text("theta_edge_level").notNull(),
  regime: text("regime").notNull(),
  historicalSpikeRate: real("historical_spike_rate").notNull(),
  breakoutUpper: real("breakout_upper").notNull(),
  breakoutLower: real("breakout_lower").notNull(),
  environmentState: text("environment_state").notNull(),
  breakoutNext: jsonb("breakout_next").notNull(),
  gammaWindows: jsonb("gamma_windows").notNull(),
  generatedAt: timestamp("generated_at").notNull().defaultNow(),
});

export const insertBuyerDataSchema = createInsertSchema(buyerData).omit({ id: true, generatedAt: true });
export type InsertBuyerData = z.infer<typeof insertBuyerDataSchema>;
export type BuyerData = typeof buyerData.$inferSelect;

// Runtime validation schemas (for API responses - these validate the JSON structure)
export const marketDataSchema = z.object({
  spot: z.number(),
  spot_change: z.number(),
  vix: z.number(),
  vix_change: z.number(),
  regime: z.enum(["LOW_VOL_BULL", "HIGH_VOL_BULL", "LOW_VOL_BEAR", "HIGH_VOL_BEAR", "SIDEWAYS", "CHOPPY"]),
});

export const horizonDataSchema = z.object({
  direction: z.enum(["UP", "DOWN", "NEUTRAL"]),
  expected_move_points: z.number(),
  conviction: z.number().min(0).max(1),
});

export const directionDataSchema = z.object({
  today: horizonDataSchema,
  horizons: z.object({
    t1: horizonDataSchema,
    t3: horizonDataSchema,
    t5: horizonDataSchema,
    t10: horizonDataSchema,
    t20: horizonDataSchema,
    t40: horizonDataSchema,
  }),
  risk_score: z.number().min(0).max(1),
});

export const rangeSchema = z.object({
  lower: z.number(),
  upper: z.number(),
});

export const sellerDataResponseSchema = z.object({
  safe_range: rangeSchema,
  max_pain: rangeSchema,
  expiry_stress: z.object({
    score: z.number().min(0).max(100),
    level: z.enum(["LOW", "MEDIUM", "HIGH"]),
  }),
  trap: z.object({
    score: z.number().min(0).max(100),
    level: z.enum(["LOW", "MEDIUM", "HIGH"]),
  }),
  skew: z.object({
    put_skew: z.number(),
    call_skew: z.number(),
  }),
  regime: z.enum(["CALM", "CAUTIOUS", "HOSTILE"]),
  breach_probabilities: z.array(z.object({
    distance: z.number(),
    probability: z.number().min(0).max(1),
  })),
  historical_hit_rate: z.number().min(0).max(1),
  daily_flag: z.enum(["FAVOURABLE", "NEUTRAL", "RISKY"]),
});

export const buyerDataResponseSchema = z.object({
  breakout_today: z.object({
    score: z.number().min(0).max(100),
    level: z.enum(["LOW", "MODERATE", "HIGH"]),
  }),
  spike_direction_bias: z.object({
    up_prob: z.number().min(0).max(1),
    down_prob: z.number().min(0).max(1),
  }),
  theta_edge: z.object({
    score: z.number().min(0).max(100),
    level: z.enum(["POOR", "FAIR", "GOOD"]),
  }),
  breakout_next: z.array(z.object({
    day_offset: z.number(),
    score: z.number().min(0).max(100),
  })),
  gamma_windows: z.array(z.object({
    time_block: z.string(),
    score: z.number().min(0).max(100),
  })),
  regime: z.enum(["TREND_FOLLOWING", "MEAN_REVERTING", "CHOPPY"]),
  historical_spike_rate: z.number().min(0).max(1),
  breakout_levels: z.object({
    upper: z.number(),
    lower: z.number(),
  }),
  environment_state: z.enum(["PREMIUM_FRIENDLY", "CAUTIOUS", "AVOID_FULL_RISK"]),
});

export const aegisMatrixDataSchema = z.object({
  generated_at: z.string(),
  market: marketDataSchema,
  direction: directionDataSchema,
  seller: sellerDataResponseSchema,
  buyer: buyerDataResponseSchema,
});

// Type exports for runtime use
export type HorizonData = z.infer<typeof horizonDataSchema>;
export type DirectionData = z.infer<typeof directionDataSchema>;
export type SellerDataResponse = z.infer<typeof sellerDataResponseSchema>;
export type BuyerDataResponse = z.infer<typeof buyerDataResponseSchema>;
export type AegisMatrixData = z.infer<typeof aegisMatrixDataSchema>;

File: schema.py
"""
JSON schema validation for aegismatrix.json using Pydantic.
Ensures frontend never breaks due to missing or malformed fields.
"""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Literal
from datetime import datetime


# Direction Block Models
class DirectionHorizon(BaseModel):
    label: str = Field(..., description="Human-readable horizon (e.g., 'Tomorrow')")
    direction: Literal["UP", "DOWN", "NEUTRAL"]
    expected_move_points: float = Field(..., description="Expected move in points (signed)")
    conviction: float = Field(..., ge=0, le=1, description="Confidence 0-1")


class DirectionToday(BaseModel):
    direction: Literal["UP", "DOWN", "NEUTRAL"]
    expected_move_points: float
    conviction: float = Field(..., ge=0, le=1)
    intraday_volatility_score: float = Field(..., ge=0, le=1)
    last_update: str = Field(..., description="ISO8601 with timezone")


class DirectionBlock(BaseModel):
    today: DirectionToday
    horizons: Dict[str, DirectionHorizon]
    risk_score: float = Field(..., ge=0, le=1)


# Market Block
class MarketBlock(BaseModel):
    spot: float
    spot_change: float
    spot_change_pct: float
    vix: float
    vix_change: float
    vix_change_pct: float
    regime: Literal["LOW_VOL_BULL", "HIGH_VOL_BEAR", "SIDEWAYS_CHOP", "TREND_FOLLOWING"]


# Seller Block Models
class SafeRange(BaseModel):
    lower: float
    upper: float
    horizon_days: int


class MaxPain(BaseModel):
    lower: float
    upper: float
    confidence: float = Field(..., ge=0, le=1)


class TrapRisk(BaseModel):
    score: float = Field(..., ge=0, le=1)
    label: Literal["LOW", "MEDIUM", "HIGH"]
    iv_percentile: float = Field(..., ge=0, le=1)
    rv_percentile: float = Field(..., ge=0, le=1)


class Skew(BaseModel):
    put_skew: float = Field(..., ge=-1, le=1)
    call_skew: float = Field(..., ge=-1, le=1)
    net_skew: float = Field(..., ge=-1, le=1)


class ExpiryStress(BaseModel):
    score: float = Field(..., ge=0, le=1)
    label: Literal["CALM", "CAUTION", "HOSTILE"]


class BreachProbability(BaseModel):
    distance: int = Field(..., gt=0, description="Points away from spot")
    probability: float = Field(..., ge=0, le=1)


class SellerFlag(BaseModel):
    label: Literal["FAVOURABLE", "CAUTION", "HOSTILE"]
    color: Literal["GREEN", "AMBER", "RED"]
    reasons: List[str]


class SellerBlock(BaseModel):
    safe_range: SafeRange
    max_pain: MaxPain
    trap: TrapRisk
    skew: Skew
    expiry_stress: ExpiryStress
    breach_probabilities: List[BreachProbability]
    seller_flag: SellerFlag


# Buyer Block Models
class BuyerScore(BaseModel):
    score: float = Field(..., ge=0, le=1)
    label: Literal["LOW", "MEDIUM", "HIGH"]


class BreakoutNext(BaseModel):
    day_offset: int = Field(..., ge=1, le=5)
    score: float = Field(..., ge=0, le=1)
    label: Literal["LOW", "MEDIUM", "HIGH"]


class SpikeDirectionBias(BaseModel):
    up_prob: float = Field(..., ge=0, le=1)
    down_prob: float = Field(..., ge=0, le=1)


class GammaWindow(BaseModel):
    window: str = Field(..., description="HH:MM-HH:MM format")
    score: float = Field(..., ge=0, le=1)


class ThetaEdge(BaseModel):
    score: float = Field(..., ge=0, le=1)
    label: Literal["DONT_WASTE_PREMIUM", "BORDERLINE", "EDGE_JUSTIFIES_PREMIUM"]


class BuyerEnvironment(BaseModel):
    label: Literal["PREMIUM_FRIENDLY", "SPECULATIVE_ONLY", "UNFAVOURABLE"]
    color: Literal["GREEN", "AMBER", "RED"]
    reasons: List[str]


class BuyerBlock(BaseModel):
    breakout_today: BuyerScore
    breakout_next: List[BreakoutNext]
    spike_direction_bias: SpikeDirectionBias
    gamma_windows: List[GammaWindow]
    theta_edge: ThetaEdge
    regime: Literal["TREND_FOLLOWING", "MEAN_REVERT", "CHOPPY"]
    buyer_environment: BuyerEnvironment


# Top-level schema
class AegisMatrixPayload(BaseModel):
    generated_at: str = Field(..., description="ISO8601 UTC string")
    market: MarketBlock
    direction: DirectionBlock
    seller: SellerBlock
    buyer: BuyerBlock


def validate_payload(payload: dict) -> None:
    """
    Validate payload against schema. Raises ValidationError if invalid.
    
    Args:
        payload: Dict to validate
        
    Raises:
        pydantic.ValidationError
    """
    AegisMatrixPayload(**payload)
